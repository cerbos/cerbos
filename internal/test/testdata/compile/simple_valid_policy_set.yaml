---
mainDef: "resource_policies/leave_request_20210210.yaml"
inputDefs:
  "resource_policies/leave_request_20210210.yaml":
    apiVersion: api.cerbos.dev/v1
    resourcePolicy:
      resource: leave_request
      version: "20210210"
      importDerivedRoles:
        - my_derived_roles
      rules:
        - actions: ['*']
          effect: EFFECT_ALLOW
          roles:
            - admin
        - actions: ["create"]
          derivedRoles:
            - employee_that_owns_the_record
          effect: EFFECT_ALLOW
        - actions: ["view:*"]
          derivedRoles:
            - employee_that_owns_the_record
            - direct_manager
          effect: EFFECT_ALLOW
        - actions: ["view:public"]
          derivedRoles:
            - any_employee
          effect: EFFECT_ALLOW
        - actions: ["approve"]
          condition:
            match:
              expr: request.resource.attr.status == "PENDING_APPROVAL"
          derivedRoles:
            - direct_manager
          effect: EFFECT_ALLOW

  "derived_roles/my_derived_roles.yaml":
    apiVersion: "api.cerbos.dev/v1"
    derivedRoles:
      name: my_derived_roles
      definitions:
        - name: tester
          parentRoles: ["dev", "qa"]

        - name: employee_that_owns_the_record
          parentRoles: ["employee"]
          condition:
            script: |-
              input.resource.attr.owner == input.principal.id

        - name: any_employee
          parentRoles: ["employee"]

        - name: direct_manager
          parentRoles: ["manager"]
          condition:
            match:
              all:
                of:
                  - expr: "request.resource.attr.geography == request.principal.attr.geography"
                  - expr: "request.resource.attr.geography == request.principal.attr.managed_geographies"
