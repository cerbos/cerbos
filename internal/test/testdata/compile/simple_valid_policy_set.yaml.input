-- resource_policies/leave_request_20210210.yaml --
---
apiVersion: api.cerbos.dev/v1
resourcePolicy:
  resource: leave_request
  version: "20210210"
  importDerivedRoles:
    - my_derived_roles
  schemas:
    principalSchema:
      ref: cerbos:///complex_object.json
    resourceSchema:
      ref: cerbos:///complex_object.json
  rules:
    - actions: ['*']
      effect: EFFECT_ALLOW
      roles:
        - admin
      name: wildcard
      output:
        when:
          ruleActivated: |-
            "wildcard:%s".format([request.principal.id])
          conditionNotMet: |-
            "condition_not_met:wildcard:%s".format([request.principal.id])
    - actions: ["create"]
      derivedRoles:
        - employee_that_owns_the_record
      effect: EFFECT_ALLOW
      output:
        when:
          ruleActivated: |-
            "create:%s".format([request.principal.id])
    - actions: ["view:*"]
      derivedRoles:
        - employee_that_owns_the_record
        - direct_manager
      effect: EFFECT_ALLOW
      output:
        when:
          conditionNotMet: |-
            "something_arbitrary"
    - actions: ["view:public"]
      derivedRoles:
        - any_employee
      effect: EFFECT_ALLOW
    - actions: ["approve"]
      condition:
        match:
          expr: request.resource.attr.status == "PENDING_APPROVAL"
      derivedRoles:
        - direct_manager
      effect: EFFECT_ALLOW
      output:
        when:
          ruleActivated: |-
            "pending_approval:%s".format([request.resource.attr.status == "PENDING_APPROVAL"])

-- derived_roles/my_derived_roles.yaml --
---
apiVersion: "api.cerbos.dev/v1"
derivedRoles:
  name: my_derived_roles
  definitions:
    - name: tester
      parentRoles: ["dev", "qa"]

    - name: employee_that_owns_the_record
      parentRoles: ["employee"]
      condition:
        match:
          expr: R.attr.owner == P.id

    - name: any_employee
      parentRoles: ["employee"]

    - name: direct_manager
      parentRoles: ["manager"]
      condition:
        match:
          all:
            of:
              - expr: "request.resource.attr.geography == request.principal.attr.geography"
              - expr: "request.resource.attr.geography == request.principal.attr.managed_geographies"
