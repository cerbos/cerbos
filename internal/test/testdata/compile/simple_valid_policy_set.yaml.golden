{
  "fqn": "cerbos.resource.leave_request.v20210210",
  "resourcePolicy": {
    "meta": {
      "fqn": "cerbos.resource.leave_request.v20210210",
      "resource": "leave_request",
      "version": "20210210"
    },
    "policies": [
      {
        "derivedRoles": {
          "any_employee": {
            "name": "any_employee",
            "parentRoles": {
              "employee": {}
            }
          },
          "direct_manager": {
            "name": "direct_manager",
            "parentRoles": {
              "manager": {}
            },
            "condition": {
              "all": {
                "expr": [
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.geography",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.CheckInput"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.CheckInput"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            68
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  },
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.managed_geographies",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.CheckInput"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.CheckInput"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            78
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "managed_geographies"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  }
                ]
              }
            }
          },
          "employee_that_owns_the_record": {
            "name": "employee_that_owns_the_record",
            "parentRoles": {
              "employee": {}
            },
            "condition": {
              "expr": {
                "original": "R.attr.owner == P.id",
                "checked": {
                  "referenceMap": {
                    "1": {
                      "name": "R"
                    },
                    "4": {
                      "overloadId": [
                        "equals"
                      ]
                    },
                    "5": {
                      "name": "P"
                    }
                  },
                  "typeMap": {
                    "1": {
                      "messageType": "cerbos.engine.v1.Resource"
                    },
                    "2": {
                      "mapType": {
                        "keyType": {
                          "primitive": "STRING"
                        },
                        "valueType": {
                          "dyn": {}
                        }
                      }
                    },
                    "3": {
                      "dyn": {}
                    },
                    "4": {
                      "primitive": "BOOL"
                    },
                    "5": {
                      "messageType": "cerbos.engine.v1.Principal"
                    },
                    "6": {
                      "primitive": "STRING"
                    }
                  },
                  "sourceInfo": {
                    "location": "<input>",
                    "lineOffsets": [
                      21
                    ],
                    "positions": {
                      "1": 0,
                      "2": 1,
                      "3": 6,
                      "4": 13,
                      "5": 16,
                      "6": 17
                    }
                  },
                  "expr": {
                    "id": "4",
                    "callExpr": {
                      "function": "_==_",
                      "args": [
                        {
                          "id": "3",
                          "selectExpr": {
                            "operand": {
                              "id": "2",
                              "selectExpr": {
                                "operand": {
                                  "id": "1",
                                  "identExpr": {
                                    "name": "R"
                                  }
                                },
                                "field": "attr"
                              }
                            },
                            "field": "owner"
                          }
                        },
                        {
                          "id": "6",
                          "selectExpr": {
                            "operand": {
                              "id": "5",
                              "identExpr": {
                                "name": "P"
                              }
                            },
                            "field": "id"
                          }
                        }
                      ]
                    }
                  }
                }
              }
            }
          }
        },
        "rules": [
          {
            "name": "wildcard",
            "actions": {
              "*": {}
            },
            "roles": {
              "admin": {}
            },
            "effect": "EFFECT_ALLOW",
            "output": {
              "original": "\"wildcard:%s\".format([request.principal.id])",
              "checked": {
                "referenceMap": {
                  "2": {
                    "overloadId": [
                      "string_format"
                    ]
                  },
                  "4": {
                    "name": "request"
                  }
                },
                "typeMap": {
                  "1": {
                    "primitive": "STRING"
                  },
                  "2": {
                    "primitive": "STRING"
                  },
                  "3": {
                    "listType": {
                      "elemType": {
                        "primitive": "STRING"
                      }
                    }
                  },
                  "4": {
                    "messageType": "cerbos.engine.v1.CheckInput"
                  },
                  "5": {
                    "messageType": "cerbos.engine.v1.Principal"
                  },
                  "6": {
                    "primitive": "STRING"
                  }
                },
                "sourceInfo": {
                  "location": "<input>",
                  "lineOffsets": [
                    45
                  ],
                  "positions": {
                    "1": 0,
                    "2": 20,
                    "3": 21,
                    "4": 22,
                    "5": 29,
                    "6": 39
                  }
                },
                "expr": {
                  "id": "2",
                  "callExpr": {
                    "target": {
                      "id": "1",
                      "constExpr": {
                        "stringValue": "wildcard:%s"
                      }
                    },
                    "function": "format",
                    "args": [
                      {
                        "id": "3",
                        "listExpr": {
                          "elements": [
                            {
                              "id": "6",
                              "selectExpr": {
                                "operand": {
                                  "id": "5",
                                  "selectExpr": {
                                    "operand": {
                                      "id": "4",
                                      "identExpr": {
                                        "name": "request"
                                      }
                                    },
                                    "field": "principal"
                                  }
                                },
                                "field": "id"
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              }
            }
          },
          {
            "name": "rule-002",
            "actions": {
              "create": {}
            },
            "derivedRoles": {
              "employee_that_owns_the_record": {}
            },
            "effect": "EFFECT_ALLOW",
            "output": {
              "original": "\"create:%s\".format([request.principal.id])",
              "checked": {
                "referenceMap": {
                  "2": {
                    "overloadId": [
                      "string_format"
                    ]
                  },
                  "4": {
                    "name": "request"
                  }
                },
                "typeMap": {
                  "1": {
                    "primitive": "STRING"
                  },
                  "2": {
                    "primitive": "STRING"
                  },
                  "3": {
                    "listType": {
                      "elemType": {
                        "primitive": "STRING"
                      }
                    }
                  },
                  "4": {
                    "messageType": "cerbos.engine.v1.CheckInput"
                  },
                  "5": {
                    "messageType": "cerbos.engine.v1.Principal"
                  },
                  "6": {
                    "primitive": "STRING"
                  }
                },
                "sourceInfo": {
                  "location": "<input>",
                  "lineOffsets": [
                    43
                  ],
                  "positions": {
                    "1": 0,
                    "2": 18,
                    "3": 19,
                    "4": 20,
                    "5": 27,
                    "6": 37
                  }
                },
                "expr": {
                  "id": "2",
                  "callExpr": {
                    "target": {
                      "id": "1",
                      "constExpr": {
                        "stringValue": "create:%s"
                      }
                    },
                    "function": "format",
                    "args": [
                      {
                        "id": "3",
                        "listExpr": {
                          "elements": [
                            {
                              "id": "6",
                              "selectExpr": {
                                "operand": {
                                  "id": "5",
                                  "selectExpr": {
                                    "operand": {
                                      "id": "4",
                                      "identExpr": {
                                        "name": "request"
                                      }
                                    },
                                    "field": "principal"
                                  }
                                },
                                "field": "id"
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              }
            }
          },
          {
            "name": "rule-003",
            "actions": {
              "view:*": {}
            },
            "derivedRoles": {
              "direct_manager": {},
              "employee_that_owns_the_record": {}
            },
            "effect": "EFFECT_ALLOW",
            "output": {
              "original": "\"something_arbitrary\"",
              "checked": {
                "typeMap": {
                  "1": {
                    "primitive": "STRING"
                  }
                },
                "sourceInfo": {
                  "location": "<input>",
                  "lineOffsets": [
                    22
                  ],
                  "positions": {
                    "1": 0
                  }
                },
                "expr": {
                  "id": "1",
                  "constExpr": {
                    "stringValue": "something_arbitrary"
                  }
                }
              }
            }
          },
          {
            "name": "rule-004",
            "actions": {
              "view:public": {}
            },
            "derivedRoles": {
              "any_employee": {}
            },
            "effect": "EFFECT_ALLOW"
          },
          {
            "name": "rule-005",
            "actions": {
              "approve": {}
            },
            "derivedRoles": {
              "direct_manager": {}
            },
            "condition": {
              "expr": {
                "original": "request.resource.attr.status == \"PENDING_APPROVAL\"",
                "checked": {
                  "referenceMap": {
                    "1": {
                      "name": "request"
                    },
                    "5": {
                      "overloadId": [
                        "equals"
                      ]
                    }
                  },
                  "typeMap": {
                    "1": {
                      "messageType": "cerbos.engine.v1.CheckInput"
                    },
                    "2": {
                      "messageType": "cerbos.engine.v1.Resource"
                    },
                    "3": {
                      "mapType": {
                        "keyType": {
                          "primitive": "STRING"
                        },
                        "valueType": {
                          "dyn": {}
                        }
                      }
                    },
                    "4": {
                      "dyn": {}
                    },
                    "5": {
                      "primitive": "BOOL"
                    },
                    "6": {
                      "primitive": "STRING"
                    }
                  },
                  "sourceInfo": {
                    "location": "<input>",
                    "lineOffsets": [
                      51
                    ],
                    "positions": {
                      "1": 0,
                      "2": 7,
                      "3": 16,
                      "4": 21,
                      "5": 29,
                      "6": 32
                    }
                  },
                  "expr": {
                    "id": "5",
                    "callExpr": {
                      "function": "_==_",
                      "args": [
                        {
                          "id": "4",
                          "selectExpr": {
                            "operand": {
                              "id": "3",
                              "selectExpr": {
                                "operand": {
                                  "id": "2",
                                  "selectExpr": {
                                    "operand": {
                                      "id": "1",
                                      "identExpr": {
                                        "name": "request"
                                      }
                                    },
                                    "field": "resource"
                                  }
                                },
                                "field": "attr"
                              }
                            },
                            "field": "status"
                          }
                        },
                        {
                          "id": "6",
                          "constExpr": {
                            "stringValue": "PENDING_APPROVAL"
                          }
                        }
                      ]
                    }
                  }
                }
              }
            },
            "effect": "EFFECT_ALLOW",
            "output": {
              "original": "\"pending_approval:%s\".format([request.resource.attr.status == \"PENDING_APPROVAL\"])",
              "checked": {
                "referenceMap": {
                  "2": {
                    "overloadId": [
                      "string_format"
                    ]
                  },
                  "4": {
                    "name": "request"
                  },
                  "8": {
                    "overloadId": [
                      "equals"
                    ]
                  }
                },
                "typeMap": {
                  "1": {
                    "primitive": "STRING"
                  },
                  "2": {
                    "primitive": "STRING"
                  },
                  "3": {
                    "listType": {
                      "elemType": {
                        "primitive": "BOOL"
                      }
                    }
                  },
                  "4": {
                    "messageType": "cerbos.engine.v1.CheckInput"
                  },
                  "5": {
                    "messageType": "cerbos.engine.v1.Resource"
                  },
                  "6": {
                    "mapType": {
                      "keyType": {
                        "primitive": "STRING"
                      },
                      "valueType": {
                        "dyn": {}
                      }
                    }
                  },
                  "7": {
                    "dyn": {}
                  },
                  "8": {
                    "primitive": "BOOL"
                  },
                  "9": {
                    "primitive": "STRING"
                  }
                },
                "sourceInfo": {
                  "location": "<input>",
                  "lineOffsets": [
                    83
                  ],
                  "positions": {
                    "1": 0,
                    "2": 28,
                    "3": 29,
                    "4": 30,
                    "5": 37,
                    "6": 46,
                    "7": 51,
                    "8": 59,
                    "9": 62
                  }
                },
                "expr": {
                  "id": "2",
                  "callExpr": {
                    "target": {
                      "id": "1",
                      "constExpr": {
                        "stringValue": "pending_approval:%s"
                      }
                    },
                    "function": "format",
                    "args": [
                      {
                        "id": "3",
                        "listExpr": {
                          "elements": [
                            {
                              "id": "8",
                              "callExpr": {
                                "function": "_==_",
                                "args": [
                                  {
                                    "id": "7",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "6",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "5",
                                            "selectExpr": {
                                              "operand": {
                                                "id": "4",
                                                "identExpr": {
                                                  "name": "request"
                                                }
                                              },
                                              "field": "resource"
                                            }
                                          },
                                          "field": "attr"
                                        }
                                      },
                                      "field": "status"
                                    }
                                  },
                                  {
                                    "id": "9",
                                    "constExpr": {
                                      "stringValue": "PENDING_APPROVAL"
                                    }
                                  }
                                ]
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              }
            }
          }
        ],
        "schemas": {
          "principalSchema": {
            "ref": "cerbos:///complex_object.json"
          },
          "resourceSchema": {
            "ref": "cerbos:///complex_object.json"
          }
        }
      }
    ],
    "schemas": {
      "principalSchema": {
        "ref": "cerbos:///complex_object.json"
      },
      "resourceSchema": {
        "ref": "cerbos:///complex_object.json"
      }
    }
  }
}